// arbundles/build/web/cjs/src/signing/chains/arconnectSigner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../../constants");
const base64url_1 = __importDefault(require("base64url"));
const utils_1 = require("../../nodeUtils.js");
class InjectedArweaveSigner {
    constructor(windowArweaveWallet, arweave) {
        this.ownerLength = constants_1.SIG_CONFIG[constants_1.SignatureConfig.ARWEAVE].pubLength;
        this.signatureLength = constants_1.SIG_CONFIG[constants_1.SignatureConfig.ARWEAVE].sigLength;
        this.signatureType = constants_1.SignatureConfig.ARWEAVE;
        this.signer = windowArweaveWallet;
        this.arweave = arweave;
    }
    setPublicKey() {
        return __awaiter(this, void 0, void 0, function* () {
            const arOwner = yield this.signer.getActivePublicKey();
            this.publicKey = base64url_1.default.toBuffer(arOwner);
        });
    }
    sign(message) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.publicKey) {
                yield this.setPublicKey();
            }
            const algorithm = {
                name: "RSA-PSS",
                saltLength: 32,
            };
            const signature = yield this.signer.signature(message, algorithm);
            const buf = new Uint8Array(Object.values(signature).map((v) => +v));
            return buf;
        });
    }
    static verify(pk, message, signature) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.getCryptoDriver)().verify(pk, message, signature);
        });
    }
}
exports.default = InjectedArweaveSigner;
