// arbundles/build/web/cjs/src/signing/chains/injectedEthereumSigner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InjectedEthereumSigner = void 0;
const hash_1 = require("@ethersproject/hash");
const signing_key_1 = require("@ethersproject/signing-key");
const constants_1 = require("../../constants");
const bytes_1 = require("@ethersproject/bytes");
const transactions_1 = require("@ethersproject/transactions");
const wallet_1 = require("@ethersproject/wallet");
class InjectedEthereumSigner {
    constructor(provider) {
        this.ownerLength = constants_1.SIG_CONFIG[constants_1.SignatureConfig.ETHEREUM].pubLength;
        this.signatureLength = constants_1.SIG_CONFIG[constants_1.SignatureConfig.ETHEREUM].sigLength;
        this.signatureType = constants_1.SignatureConfig.ETHEREUM;
        this.signer = provider.getSigner();
    }
    setPublicKey() {
        return __awaiter(this, void 0, void 0, function* () {
            const address = "sign this message to connect to Bundlr.Network";
            const signedMsg = yield this.signer.signMessage(address);
            const hash = yield (0, hash_1.hashMessage)(address);
            const recoveredKey = (0, signing_key_1.recoverPublicKey)((0, bytes_1.arrayify)(hash), signedMsg);
            this.publicKey = Buffer.from((0, bytes_1.arrayify)(recoveredKey));
        });
    }
    sign(message) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.publicKey) {
                yield this.setPublicKey();
            }
            const sig = yield this.signer.signMessage(message);
            return Buffer.from(sig.slice(2), "hex");
        });
    }
    static verify(pk, message, signature) {
        const address = (0, transactions_1.computeAddress)(pk);
        return (0, wallet_1.verifyMessage)(message, signature) === address;
    }
}
exports.InjectedEthereumSigner = InjectedEthereumSigner;
exports.default = InjectedEthereumSigner;
