// arbundles/build/node/cjs/src/file/FileBundle.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileBundle = void 0;
const FileDataItem_1 = __importDefault(require("./FileDataItem"));
const fs_1 = require("fs");
const utils_1 = require("../utils");
const fs_2 = require("fs");
const multistream_1 = __importDefault(require("multistream"));
const util_1 = require("util");
const base64url_1 = __importDefault(require("base64url"));
const promises_1 = require("stream/promises");
const path_1 = require("path");
// import { Readable } from 'stream';
// import { createTransactionAsync } from 'arweave-stream';
// import { pipeline } from 'stream/promises';
const read = (0, util_1.promisify)(fs_2.read);
class FileBundle {
    constructor(headerFile, txs) {
        this.headerFile = headerFile;
        this.txs = txs;
    }
    static fromDir(dir) {
        return __awaiter(this, void 0, void 0, function* () {
            const txs = [];
            for (const p of yield fs_1.promises.readdir(dir)) {
                const fullPath = (0, path_1.resolve)(dir, p);
                // if it's an item (not a dir,not the header file, actually exists in FS) add to txs array
                if (p !== "header" &&
                    (yield fs_1.promises
                        .stat(fullPath)
                        .then((e) => !e.isDirectory())
                        .catch((_) => false)))
                    txs.push(fullPath);
            }
            return new FileBundle(dir + "/header", txs);
        });
    }
    length() {
        return __awaiter(this, void 0, void 0, function* () {
            const handle = yield fs_1.promises.open(this.headerFile, "r");
            const lengthBuffer = yield read(handle.fd, Buffer.allocUnsafe(32), 0, 32, 0).then((r) => r.buffer);
            yield handle.close();
            return (0, utils_1.byteArrayToLong)(lengthBuffer);
        });
    }
    get items() {
        return this.itemsGenerator();
    }
    get(index) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof index === "number") {
                if (index > (yield this.length())) {
                    throw new RangeError("Index out of range");
                }
                return this.getByIndex(index);
            }
            else {
                return this.getById(index);
            }
        });
    }
    getIds() {
        var _a, e_1, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            const ids = new Array(yield this.length());
            let count = 0;
            try {
                for (var _d = true, _e = __asyncValues(this.getHeaders()), _f; _f = yield _e.next(), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const { id } = _c;
                        ids[count] = id;
                        count++;
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield _b.call(_e);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return ids;
        });
    }
    getRaw() {
        var _a, e_2, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            const streams = [(0, fs_1.createReadStream)(this.headerFile), ...this.txs.map((t) => (0, fs_1.createReadStream)(t))];
            const stream = multistream_1.default.obj(streams);
            let buff = Buffer.allocUnsafe(0);
            try {
                for (var _d = true, stream_1 = __asyncValues(stream), stream_1_1; stream_1_1 = yield stream_1.next(), _a = stream_1_1.done, !_a;) {
                    _c = stream_1_1.value;
                    _d = false;
                    try {
                        const chunk = _c;
                        buff = Buffer.concat([buff, Buffer.from(chunk)]);
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = stream_1.return)) yield _b.call(stream_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return buff;
        });
    }
    toTransaction(attributes, arweave, jwk) {
        return __awaiter(this, void 0, void 0, function* () {
            const streams = [(0, fs_1.createReadStream)(this.headerFile), ...this.txs.map((t) => (0, fs_1.createReadStream)(t))];
            const stream = multistream_1.default.obj(streams);
            const tx = yield (0, promises_1.pipeline)(stream, arweave.stream.createTransactionAsync(attributes, jwk));
            tx.addTag("Bundle-Format", "binary");
            tx.addTag("Bundle-Version", "2.0.0");
            return tx;
        });
    }
    signAndSubmit(arweave, jwk, tags = []) {
        return __awaiter(this, void 0, void 0, function* () {
            const tx = yield this.toTransaction({}, arweave, jwk);
            // tx.addTag("Bundle-Format", "binary");
            // tx.addTag("Bundle-Version", "2.0.0");
            for (const { name, value } of tags) {
                tx.addTag(name, value);
            }
            yield arweave.transactions.sign(tx, jwk);
            const streams2 = [(0, fs_1.createReadStream)(this.headerFile), ...this.txs.map((t) => (0, fs_1.createReadStream)(t))];
            const stream2 = multistream_1.default.obj(streams2);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            yield (0, promises_1.pipeline)(stream2, arweave.stream.uploadTransactionAsync(tx, true));
            return tx;
        });
    }
    getHeaders() {
        return __asyncGenerator(this, arguments, function* getHeaders_1() {
            const handle = yield __await(fs_1.promises.open(this.headerFile, "r"));
            for (let i = 32; i < 32 + 64 * (yield __await(this.length())); i += 64) {
                yield yield __await({
                    offset: (0, utils_1.byteArrayToLong)(yield __await(read(handle.fd, Buffer.allocUnsafe(32), 0, 32, i).then((r) => r.buffer))),
                    id: yield __await(read(handle.fd, Buffer.allocUnsafe(32), 0, 32, i + 32).then((r) => base64url_1.default.encode(r.buffer))),
                });
            }
            yield __await(handle.close());
        });
    }
    itemsGenerator() {
        return __asyncGenerator(this, arguments, function* itemsGenerator_1() {
            var _a, e_3, _b, _c;
            let counter = 0;
            try {
                for (var _d = true, _e = __asyncValues(this.getHeaders()), _f; _f = yield __await(_e.next()), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const { id } = _c;
                        yield yield __await(new FileDataItem_1.default(this.txs[counter], base64url_1.default.toBuffer(id)));
                        counter++;
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_3) throw e_3.error; }
            }
        });
    }
    getById(txId) {
        var _a, e_4, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            let counter = 0;
            try {
                for (var _d = true, _e = __asyncValues(this.getHeaders()), _f; _f = yield _e.next(), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const { id } = _c;
                        if (id === txId)
                            return new FileDataItem_1.default(this.txs[counter], base64url_1.default.toBuffer(id));
                        counter++;
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield _b.call(_e);
                }
                finally { if (e_4) throw e_4.error; }
            }
            throw new Error("Can't find by id");
        });
    }
    getByIndex(index) {
        var _a, e_5, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            let count = 0;
            try {
                for (var _d = true, _e = __asyncValues(this.getHeaders()), _f; _f = yield _e.next(), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const { id } = _c;
                        if (count === index) {
                            return new FileDataItem_1.default(this.txs[count], base64url_1.default.toBuffer(id));
                        }
                        count++;
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield _b.call(_e);
                }
                finally { if (e_5) throw e_5.error; }
            }
            throw new Error("Can't find by index");
        });
    }
}
exports.FileBundle = FileBundle;
exports.default = FileBundle;
