// arbundles/build/node/cjs/src/nodeUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCryptoDriver = exports.CryptoDriver = exports.Arweave = exports.deepHash = exports.Transaction = exports.concatBuffers = exports.stringToBuffer = void 0;
const crypto_1 = require("crypto");
const node_driver_1 = __importDefault(require("@irys/arweave/node/node-driver"));
// import CryptoInterface from "arweave/node/lib/crypto/crypto-interface";
var utils_1 = require("@irys/arweave/common/lib/utils");
Object.defineProperty(exports, "stringToBuffer", { enumerable: true, get: function () { return utils_1.stringToBuffer; } });
Object.defineProperty(exports, "concatBuffers", { enumerable: true, get: function () { return utils_1.concatBuffers; } });
var transaction_1 = require("@irys/arweave/common/lib/transaction");
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return __importDefault(transaction_1).default; } });
var deepHash_1 = require("./deepHash");
Object.defineProperty(exports, "deepHash", { enumerable: true, get: function () { return deepHash_1.deepHash; } });
var node_1 = require("@irys/arweave/node");
Object.defineProperty(exports, "Arweave", { enumerable: true, get: function () { return __importDefault(node_1).default; } });
// hack as ESM won't unpack .default CJS imports, so we do so dynamically
// eslint-disable-next-line @typescript-eslint/dot-notation
const driver = node_driver_1.default["default"] ? node_driver_1.default["default"] : node_driver_1.default;
class CryptoDriver extends driver {
    getPublicKey(jwk) {
        return (0, crypto_1.createPublicKey)({
            key: this.jwkToPem(jwk),
            type: "pkcs1",
            format: "pem",
        })
            .export({
            format: "pem",
            type: "pkcs1",
        })
            .toString();
    }
}
exports.CryptoDriver = CryptoDriver;
let driverInstance;
function getCryptoDriver() {
    return (driverInstance !== null && driverInstance !== void 0 ? driverInstance : (driverInstance = new CryptoDriver()));
}
exports.getCryptoDriver = getCryptoDriver;
