// @raydium-io/raydium-sdk/lib/esm/marshmallow/buffer-layout.js
import { bits as _bits, BitStructure as _BitStructure, blob as _blob, Blob as _Blob, cstr as _cstr, f32 as _f32, f32be as _f32be, f64 as _f64, f64be as _f64be, greedy as _greedy, Layout as _Layout, ns64 as _ns64, ns64be as _ns64be, nu64 as _nu64, nu64be as _nu64be, offset as _offset, s16 as _s16, s16be as _s16be, s24 as _s24, s24be as _s24be, s32 as _s32, s32be as _s32be, s40 as _s40, s40be as _s40be, s48 as _s48, s48be as _s48be, s8 as _s8, seq as _seq, struct as _struct, Structure as _Structure, u16 as _u16, u16be as _u16be, u24 as _u24, u24be as _u24be, u32 as _u32, u32be as _u32be, u40 as _u40, u40be as _u40be, u48 as _u48, u48be as _u48be, u8 as _u8, UInt as _UInt, union as _union, Union as _Union, unionLayoutDiscriminator as _unionLayoutDiscriminator, utf8 as _utf8, } from '@solana/buffer-layout';
export const Layout = _Layout;
export const Structure = _Structure;
export const Union = _Union;
export const BitStructure = _BitStructure;
export const UInt = _UInt;
export const Blob = _Blob;
//#endregion
export const greedy = _greedy;
export const u8 = _u8;
export const u16 = _u16;
export const u24 = _u24;
export const u32 = _u32;
export const u40 = _u40;
export const u48 = _u48;
export const nu64 = _nu64;
export const u16be = _u16be;
export const u24be = _u24be;
export const u32be = _u32be;
export const u40be = _u40be;
export const u48be = _u48be;
export const nu64be = _nu64be;
export const s8 = _s8;
export const s16 = _s16;
export const s24 = _s24;
export const s32 = _s32;
export const s40 = _s40;
export const s48 = _s48;
export const ns64 = _ns64;
export const s16be = _s16be;
export const s24be = _s24be;
export const s32be = _s32be;
export const s40be = _s40be;
export const s48be = _s48be;
export const ns64be = _ns64be;
export const f32 = _f32;
export const f32be = _f32be;
export const f64 = _f64;
export const f64be = _f64be;
export const struct = _struct;
export const seq = _seq;
export const union = _union;
export const unionLayoutDiscriminator = _unionLayoutDiscriminator;
export const blob = _blob;
export const cstr = _cstr;
export const utf8 = _utf8;
export const bits = _bits; // TODO: not quite sure
export const offset = _offset;
