// @irys/query/build/cjs/queries/arweave/blocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.arweaveBlocksQuery = exports.blocksVars = void 0;
const block_1 = require("./block");
// default variables
exports.blocksVars = {
    ids: undefined,
    // height: undefined,
    minHeight: undefined,
    maxHeight: undefined,
    pageSize: 10,
    after: undefined,
    sort: "DESC", // REMAPPED
};
exports.arweaveBlocksQuery = {
    name: "blocks",
    query: block_1.block,
    vars: exports.blocksVars,
    enumValues: ["sort"],
    remapVars: {
        pageSize: "first",
        // replace ASC/DESC to HEIGHT prefixed versions
        sort: (k, v) => [k, v === "ASC" ? "HEIGHT_ASC" : "HEIGHT_DESC"],
        minHeight: (_k, v, vars) => {
            vars.height = Object.assign(Object.assign({}, vars.height), { min: v });
            vars.minHeight = undefined;
        },
        maxHeight: (_k, v, vars) => {
            vars.height = Object.assign(Object.assign({}, vars.height), { max: v });
            vars.maxHeight = undefined;
        },
    },
    paging: {
        hasNextPage: "hasNextPage",
        cursor: "cursor",
    },
};
