// @irys/query/build/cjs/queries/arweave/transactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.arweaveTransactionsQuery = exports.transactionsVars = void 0;
const transaction_1 = require("./transaction");
// default variables
exports.transactionsVars = {
    ids: undefined,
    from: undefined,
    to: undefined,
    tags: undefined,
    bundledIn: undefined,
    // block: undefined,
    minHeight: undefined,
    maxHeight: undefined,
    pageSize: 10,
    after: undefined,
    sort: "DESC", // REMAPPED
};
exports.arweaveTransactionsQuery = {
    name: "transactions",
    query: transaction_1.transaction,
    enumValues: ["sort"],
    vars: exports.transactionsVars,
    remapVars: {
        pageSize: "first",
        from: "owners",
        to: "recipients",
        // replace ASC/DESC to HEIGHT prefixed versions
        sort: (k, v) => [k, v === "ASC" ? "HEIGHT_ASC" : "HEIGHT_DESC"],
        minHeight: (_k, v, vars) => {
            vars.block = Object.assign(Object.assign({}, vars.block), { min: v });
            vars.minHeight = undefined;
        },
        maxHeight: (_k, v, vars) => {
            vars.block = Object.assign(Object.assign({}, vars.block), { max: v });
            vars.maxHeight = undefined;
        },
    },
    paging: {
        hasNextPage: "hasNextPage",
        cursor: "cursor",
    },
};
