// @raydium-io/raydium-sdk/lib/cjs/marshmallow/buffer-layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.offset = exports.bits = exports.utf8 = exports.cstr = exports.blob = exports.unionLayoutDiscriminator = exports.union = exports.seq = exports.struct = exports.f64be = exports.f64 = exports.f32be = exports.f32 = exports.ns64be = exports.s48be = exports.s40be = exports.s32be = exports.s24be = exports.s16be = exports.ns64 = exports.s48 = exports.s40 = exports.s32 = exports.s24 = exports.s16 = exports.s8 = exports.nu64be = exports.u48be = exports.u40be = exports.u32be = exports.u24be = exports.u16be = exports.nu64 = exports.u48 = exports.u40 = exports.u32 = exports.u24 = exports.u16 = exports.u8 = exports.greedy = exports.Blob = exports.UInt = exports.BitStructure = exports.Union = exports.Structure = exports.Layout = void 0;
const buffer_layout_1 = require("@solana/buffer-layout");
exports.Layout = buffer_layout_1.Layout;
exports.Structure = buffer_layout_1.Structure;
exports.Union = buffer_layout_1.Union;
exports.BitStructure = buffer_layout_1.BitStructure;
exports.UInt = buffer_layout_1.UInt;
exports.Blob = buffer_layout_1.Blob;
//#endregion
exports.greedy = buffer_layout_1.greedy;
exports.u8 = buffer_layout_1.u8;
exports.u16 = buffer_layout_1.u16;
exports.u24 = buffer_layout_1.u24;
exports.u32 = buffer_layout_1.u32;
exports.u40 = buffer_layout_1.u40;
exports.u48 = buffer_layout_1.u48;
exports.nu64 = buffer_layout_1.nu64;
exports.u16be = buffer_layout_1.u16be;
exports.u24be = buffer_layout_1.u24be;
exports.u32be = buffer_layout_1.u32be;
exports.u40be = buffer_layout_1.u40be;
exports.u48be = buffer_layout_1.u48be;
exports.nu64be = buffer_layout_1.nu64be;
exports.s8 = buffer_layout_1.s8;
exports.s16 = buffer_layout_1.s16;
exports.s24 = buffer_layout_1.s24;
exports.s32 = buffer_layout_1.s32;
exports.s40 = buffer_layout_1.s40;
exports.s48 = buffer_layout_1.s48;
exports.ns64 = buffer_layout_1.ns64;
exports.s16be = buffer_layout_1.s16be;
exports.s24be = buffer_layout_1.s24be;
exports.s32be = buffer_layout_1.s32be;
exports.s40be = buffer_layout_1.s40be;
exports.s48be = buffer_layout_1.s48be;
exports.ns64be = buffer_layout_1.ns64be;
exports.f32 = buffer_layout_1.f32;
exports.f32be = buffer_layout_1.f32be;
exports.f64 = buffer_layout_1.f64;
exports.f64be = buffer_layout_1.f64be;
exports.struct = buffer_layout_1.struct;
exports.seq = buffer_layout_1.seq;
exports.union = buffer_layout_1.union;
exports.unionLayoutDiscriminator = buffer_layout_1.unionLayoutDiscriminator;
exports.blob = buffer_layout_1.blob;
exports.cstr = buffer_layout_1.cstr;
exports.utf8 = buffer_layout_1.utf8;
exports.bits = buffer_layout_1.bits; // TODO: not quite sure
exports.offset = buffer_layout_1.offset;
