// arbundles/build/web/cjs/src/webUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCryptoDriver = exports.CryptoDriver = exports.Arweave = exports.deepHash = exports.concatBuffers = exports.stringToBuffer = void 0;
const webcrypto_driver_1 = __importDefault(require("@irys/arweave/web/webcrypto-driver"));
var utils_1 = require("@irys/arweave/common/lib/utils");
Object.defineProperty(exports, "stringToBuffer", { enumerable: true, get: function () { return utils_1.stringToBuffer; } });
Object.defineProperty(exports, "concatBuffers", { enumerable: true, get: function () { return utils_1.concatBuffers; } });
var deepHash_1 = require("./deepHash");
Object.defineProperty(exports, "deepHash", { enumerable: true, get: function () { return deepHash_1.deepHash; } });
var arweave_1 = require("@irys/arweave/web/arweave");
Object.defineProperty(exports, "Arweave", { enumerable: true, get: function () { return arweave_1.Arweave; } });
// import { sha384 as SHA384 } from "sha";
// export { default as Arweave } from "arweave/web";
// import type { Hash } from "crypto";
// export const sha384 = (): Hash => SHA384("sha384");
// hack as ESM won't unpack .default CJS imports, so we do so dynamically
// eslint-disable-next-line @typescript-eslint/dot-notation
const driver = webcrypto_driver_1.default["default"] ? webcrypto_driver_1.default["default"] : webcrypto_driver_1.default;
class CryptoDriver extends driver {
    getPublicKey(_jwk) {
        throw new Error("Unimplemented");
    }
}
exports.CryptoDriver = CryptoDriver;
let driverInstance;
function getCryptoDriver() {
    return (driverInstance !== null && driverInstance !== void 0 ? driverInstance : (driverInstance = new CryptoDriver()));
}
exports.getCryptoDriver = getCryptoDriver;
